File: buildSrc/src/main/groovy/org/stepik/gradle/plugins/jetbrains/dependency/DependencyManager.java
package org.stepik.gradle.plugins.jetbrains.dependency;

import org.gradle.api.Project;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.ConfigurationContainer;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.api.publish.ivy.internal.artifact.DefaultIvyArtifact;
import org.gradle.api.publish.ivy.internal.publication.DefaultIvyConfiguration;
import org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublicationIdentity;
import org.gradle.api.publish.ivy.internal.publisher.IvyDescriptorFileGenerator;
import org.gradle.tooling.BuildException;
import org.jetbrains.annotations.NotNull;
import org.stepik.gradle.plugins.jetbrains.ProductPluginExtension;

import java.io.File;
import java.util.LinkedHashMap;
import java.util.function.Predicate;

/**
 * @author meanmail
 */
public class DependencyManager {
    @NotNull
    public static ProductDependency resolveLocal(
            @NotNull Project project,
            @NotNull ProductPluginExtension extension,
            @NotNull File localPath,
            @NotNull String productName) {
        if (!localPath.exists() || !localPath.isDirectory()) {
            throw new BuildException("Specified idePath \'" + localPath + "\' is not path to " + productName, null);
        }

        String buildNumber = extension.getVersion();

        return createCompileDependency(buildNumber, localPath, project);
    }

    @NotNull
    private static ProductDependency createCompileDependency(
            String version,
            File classesDirectory,
            Project project) {
        return new ProductDependency(version,
                version,
                classesDirectory,
                !hasKotlinDependency(project));
    }

    private static Boolean hasKotlinDependency(@NotNull Project project) {
        ConfigurationContainer configurations = project.getConfigurations();

        Predicate<Configuration> testConfiguration = (configuration) ->
                configuration.getAllDependencies().stream()
                        .filter((dependency) -> {
                            if ("org.jetbrains.kotlin".equals(dependency.getGroup())) {
                                String name = dependency.getName();
                                if ("kotlin-runtime".equals(name) || "kotlin-stdlib".equals(name) || "kotlin-reflect".equals(
                                        name)) {
                                    return true;
                                }
                            }
                            return false;
                        })
                        .count() > 0;

        Configuration runtimeConfiguration = configurations.getByName(JavaPlugin.RUNTIME_CONFIGURATION_NAME);
        Configuration compileConfiguration = configurations.getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);

        return testConfiguration.test(runtimeConfiguration) || testConfiguration.test(compileConfiguration);
    }

    public static void register(
            @NotNull Project project,
            @NotNull final ProductDependency dependency,
            @NotNull String productName) {
        productName = productName.toLowerCase();

        final File ivyFile = getOrCreateIvyXml(dependency, productName);
        project.getRepositories().ivy(repo -> {
            repo.setUrl(dependency.getClasses());
            repo.ivyPattern(ivyFile.getAbsolutePath());
            repo.artifactPattern(dependency.getClasses().getPath() + "/[artifact].[ext]");
        });

        LinkedHashMap<String, String> map = new LinkedHashMap<>(4);
        map.put("group", "com.jetbrains");
        map.put("name", productName);
        map.put("version", dependency.getVersion());
        map.put("configuration", "compile");
        project.getDependencies().add(JavaPlugin.COMPILE_CONFIGURATION_NAME, map);
    }

    private static File getOrCreateIvyXml(
            @NotNull final ProductDependency dependency,
            @NotNull final String productName) {
        File ivyFile = new File(dependency.getClasses(), dependency.getFqn(productName) + ".xml");
        if (!ivyFile.exists()) {
            final IvyDescriptorFileGenerator generator = new IvyDescriptorFileGenerator(new DefaultIvyPublicationIdentity(
                    "com.jetbrains",
                    productName,
                    dependency.getVersion()));
            generator.addConfiguration(new DefaultIvyConfiguration("default"));
            generator.addConfiguration(new DefaultIvyConfiguration("compile"));
            generator.addConfiguration(new DefaultIvyConfiguration("sources"));
            dependency.getJarFiles()
                    .forEach(jar -> generator.addArtifact(createJarCompileDependency(jar, dependency.getClasses())));

            generator.writeTo(ivyFile);
        }

        return ivyFile;
    }

    @NotNull
    private static DefaultIvyArtifact createJarCompileDependency(File file, File baseDir) {
        String relativePath = baseDir.toURI().relativize(file.toURI()).getPath();

        String name;
        if (relativePath.endsWith(".jar")) {
            name = relativePath.substring(0, relativePath.length() - "jar".length() - 1);
        } else {
            name = relativePath;
        }

        DefaultIvyArtifact artifact = new DefaultIvyArtifact(file, name, "jar", "jar", null);
        artifact.setConf("compile");
        return artifact;
    }
}

File: buildSrc/src/main/kotlin/org/stepik/gradle/plugins/common/dependency/DependencyManager.kt
package org.stepik.gradle.plugins.common.dependency

import org.gradle.api.Project
import org.gradle.api.artifacts.Configuration
import org.gradle.api.plugins.JavaPlugin.COMPILE_CONFIGURATION_NAME
import org.gradle.api.plugins.JavaPlugin.RUNTIME_CONFIGURATION_NAME
import org.gradle.api.publish.ivy.internal.artifact.DefaultIvyArtifact
import org.gradle.api.publish.ivy.internal.publication.DefaultIvyConfiguration
import org.gradle.api.publish.ivy.internal.publication.DefaultIvyPublicationIdentity
import org.gradle.api.publish.ivy.internal.publisher.IvyDescriptorFileGenerator
import org.gradle.tooling.BuildException
import org.slf4j.LoggerFactory
import org.stepik.gradle.plugins.common.BasePlugin
import org.stepik.gradle.plugins.common.ProductPluginExtension
import org.stepik.gradle.plugins.common.Utils.downloadProduct
import org.stepik.gradle.plugins.common.Utils.getArchivePath
import org.stepik.gradle.plugins.common.Utils.untgz
import org.stepik.gradle.plugins.common.Utils.unzip
import java.io.File
import java.net.URI

object DependencyManager {
    private val logger = LoggerFactory.getLogger(DependencyManager::class.java)

    val dependencies: MutableSet<ProductDependency> = mutableSetOf()

    fun resolveRemoteMaven(project: Project, extension: ProductPluginExtension): ProductDependency {
        logger.debug("Adding ${extension.productName} repository")
        project.repositories.maven { it.url = URI(extension.repository) }

        logger.debug("Adding ${extension.productName} dependency")
        val libraryType = extension.productType
        val version = extension.version
        val group = extension.productGroup
        val name = extension.productName.toLowerCase()
        val dependency = project.dependencies.create("$group:$name$libraryType:$version")
        val configuration = project.configurations.detachedConfiguration(dependency)

        val classesDirectory = getClassesDirectory(project, configuration)
        extension.idePath = classesDirectory.path
        return createCompileDependency(project, name, version, classesDirectory)
    }

    private fun getClassesDirectory(project: Project, configuration: Configuration): File {
        val zipFile = configuration.singleFile
        logger.debug("Product zip: ${zipFile.path}")
        val directoryName = zipFile.name.removeSuffix(".zip")

        val cacheDirectory = File(zipFile.parent, directoryName)
        val markerFile = File(cacheDirectory, "markerFile")
        if (!markerFile.exists()) {
            cacheDirectory.deleteRecursively()
            cacheDirectory.mkdir()
            logger.debug("Unzipping...")
            project.copy {
                it.from(project.zipTree(zipFile))
                it.into(cacheDirectory)
            }
            markerFile.createNewFile()
            logger.debug("Unzipped")
        }
        return cacheDirectory
    }

    fun resolveLocal(project: Project, extension: ProductPluginExtension): ProductDependency? {
        val idePath = extension.ideDirectory ?: return null
        if (!idePath.exists() || !idePath.isDirectory) {
            throw BuildException("Specified idePath '$idePath' is not path to ${extension.productName}", null)
        }

        return createCompileDependency(project, extension.productName.toLowerCase(), extension.version, idePath)
    }


    private fun createCompileDependency(project: Project, productName: String,
                                        version: String, classesDirectory: File): ProductDependency {
        return ProductDependency(productName, version, version,
                classesDirectory, !hasKotlinDependency(project))
    }


    private fun hasKotlinDependency(project: Project): Boolean {
        return project.configurations.let {
            it.getByName(RUNTIME_CONFIGURATION_NAME).allDependencies.union(
                    it.getByName(COMPILE_CONFIGURATION_NAME).allDependencies).any {
                ("org.jetbrains.kotlin" == it.group) && it.name in listOf("kotlin-runtime", "kotlin-stdlib", "kotlin-reflect")
            }
        }
    }

    fun register(project: Project, dependency: ProductDependency, extension: ProductPluginExtension) {
        val name = (extension.productName + extension.productType).toLowerCase()

        val ivyFile = getOrCreateIvyXml(dependency, name)
        project.repositories.ivy {
            it.setUrl(dependency.classes)
            it.ivyPattern(ivyFile.absolutePath)
            it.artifactPattern("${dependency.classes.path}/[artifact].[ext]")
        }

        val map = mapOf(
                "group" to "com.jetbrains",
                "name" to name,
                "version" to dependency.version,
                "configuration" to "compile"
        )

        project.dependencies.add(COMPILE_CONFIGURATION_NAME, map)
        dependencies.add(dependency)
    }

    private fun getOrCreateIvyXml(dependency: ProductDependency, productName: String): File {
        val ivyFile = File(dependency.classes, "${dependency.getFqn(productName)}.xml")
        if (!ivyFile.exists()) {
            val version = dependency.version
            val identity = DefaultIvyPublicationIdentity("com.jetbrains", productName, version)
            val generator = IvyDescriptorFileGenerator(identity)
            generator.addConfiguration(DefaultIvyConfiguration("default"))
            generator.addConfiguration(DefaultIvyConfiguration("compile"))
            generator.addConfiguration(DefaultIvyConfiguration("sources"))
            dependency.jarFiles.forEach {
                generator.addArtifact(createJarCompileDependency(it, dependency.classes))
            }

            generator.writeTo(ivyFile)
        }

        return ivyFile
    }


    private fun createJarCompileDependency(file: File, baseDir: File): DefaultIvyArtifact {
        val relativePath = baseDir.toURI().relativize(file.toURI()).path

        val name = relativePath.removeSuffix(".jar")

        val artifact = DefaultIvyArtifact(file, name, "jar", "jar", null)
        artifact.conf = "compile"
        return artifact
    }

    fun resolveLocalCashRepository(project: Project, basePlugin: BasePlugin,
                                   extension: ProductPluginExtension): ProductDependency? {
        val idePath = extension.ideDirectory ?: return null
        val ideArchiveType = extension.archiveType
        val productName = basePlugin.productName

        if (!idePath.exists()) {
            val archive: File = getArchivePath(project, basePlugin, extension)
            if (!archive.exists()) {
                logger.info("Download {}", extension.repository)
                println("Download ${extension.repository}")

                downloadProduct(basePlugin, extension, archive)

                if (!archive.exists()) {
                    println("$productName not loaded from ${extension.repository}")
                    logger.warn("{} not loaded from {}", productName, extension.repository)
                    return null
                }
            }

            logger.info("{} loaded", productName)
            println("$productName Loaded")

            println("Start unarchive $productName...")

            when (ideArchiveType) {
                "zip" -> unzip(archive, idePath)
                "tar.gz" -> untgz(archive, idePath)
            }

            logger.info("Unarchived $productName to $idePath")
            println("Unarchived $productName to $idePath")
        }

        return resolveLocal(project, extension)
    }

}
